Add prompt tuples

ericliuche opened the pull request:
This change adds a new `prompt` function inside `packages/airplane/internal/prompt/index.ts` that supports the same parameter format as inline task configuration. Doing so allows us to support prompt tuples

colinking commented on packages/airplane/internal/params/params_output.ts:
```diff
@@ -11,3 +14,35 @@ export type OutputParamValueByType = Omit<
+
+export const deserializeParams = <TParams extends InputParams>(
+  params: TParams,
+  paramValues: InputToOutputParamValues<TParams>
```
```suggestion
  paramValues: InputToSerializedParamValues<TParams>
```

ericliuche commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
     return paramValue.id;
   }
   if (paramValue instanceof Date) {
-    return toISOString(paramValue);
+    // TODO: Currently we incorrectly serialize `date` parameters the same as `datetime`.
+    //       `date` parameters need to instead be formatted as `YYYY-MM-DD`.
+    return paramValue.toISOString();
```
this was incorrect before. `toISOString(paramValue)` created a format our API doesn't support

ericliuche replied:
our parameter definition format doesn't accept ISO strings, trying to define a inline task date param or prompt param with options/defaults surfaces
`Error: Api: 400 - Parameters[1].default: Dates must be formatted as '2006-01-02'`

colinking replied:
Ah, that's a bug then. Here's the PR: https://github.com/airplanedev/airport/pull/3503/files

ericliuche commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+          options:
+            "options" in param && param.options != null
+              ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                serializeOptions(param.options as any, slug, optional)
```
i wasn't able to figure out how to get this type to work :(

colinking replied:
Hmm, for some reason it's tripping on the union types here:

<img width="1305" alt="CleanShot 2022-11-23 at 12 59 06" src="https://user-images.githubusercontent.com/2907397/203616942-132adfbf-7402-4e4e-959a-58da6e5e9696.png">

And reducing that to just `string | Date`:

<img width="1064" alt="CleanShot 2022-11-23 at 12 59 34" src="https://user-images.githubusercontent.com/2907397/203616985-399bacee-39f7-4096-af69-5b4974be2e30.png">

Not sure the fix, can leave as-is.

ericliuche commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+  return serializedParams;
+};
+
+const serializeOptions = <TOption extends InputParamValueByType[keyof InputParamValueByType]>(
```
before, the option values were represented as the index into the options array, why was this the case @colinking? it created odd artifacts where if you submit a prompt with options in slack, it would render the option values as the index value

colinking replied:
It was an easy way to get a unique identifier of each element. e.g. if you have a list of objects, do you have to serialize the object to get an identifier?

ericliuche commented on packages/airplane/internal/params/params_output.ts:
```diff
@@ -11,3 +17,22 @@ export type OutputParamValueByType = Omit<
+  paramValues: InputToSerializedParamValues<TParams>
+): Resolve<InputToOutputParamValues<TParams>> => {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const output = {} as Record<keyof TParams, any>;
```
wasn't able to get this type to work without `any`

ericliuche commented on packages/airplane/internal/prompt/index.test.ts:
```diff
@@ -7,6 +7,7 @@ describe("select", () => {
     const unregister = registerRuntime(
       "mock" as RuntimeKind,
       createMockRuntime({
+        // @ts-ignore
```
wasn't able to get this type working

ericliuche commented on packages/airplane/internal/prompt/index.test.ts:
```diff
@@ -60,6 +61,7 @@ describe("select", () => {
     const unregister = registerRuntime(
       "mock" as RuntimeKind,
       createMockRuntime({
+        // @ts-ignore
```
wasn't able to get this type working

colinking commented:
@ericliuche is this ready for another review? Just re-request from me

ericliuche commented on packages/airplane/internal/prompt/index.ts:
```diff
@@ -209,3 +209,18 @@ export const select = async <TOption = unknown>(
   return options[i];
 };
+
+export const prompt = async <TParams extends PromptInputParams = Record<string, never>>(
```
i wasn't able to get `airplane.prompt(...)` and `airplane.prompt.text(...)` both to work, i tried something like
```
import { prompt } from "./internal/prompt";
import * as promptFunctions from "./internal/prompt";

prompt.confirm = promptFunctions.confirm;
prompt.text = promptFunctions.text;
prompt.longText = promptFunctions.longText;
prompt.sql = promptFunctions.sql;
prompt.number = promptFunctions.number;
prompt.integer = promptFunctions.integer;
prompt.boolean = promptFunctions.boolean;
prompt.date = promptFunctions.date;
prompt.datetime = promptFunctions.datetime;
prompt.select = promptFunctions.select;
```
but when trying to use `prompt.option.select` i would get 

<img width="876" alt="Screen Shot 2022-11-23 at 10 17 35 AM" src="https://user-images.githubusercontent.com/11494140/203582554-c380a48a-0291-49d1-9e34-30b005ec99de.png">

ericliuche replied:
and for backwards compat im guessing we'd also keep around
```
packages/airplane/prompt.ts

// Supports:
// import { text } from 'airplane/prompt'
export * from "./internal/prompt";

// Supports:
// import prompt from 'airplane/prompt'
export * as default from "./internal/prompt";
```
for now?

colinking replied:
Hmm, weird. I'll TAL.

colinking replied:
Seems like you can just do:

```
export const prompt = async <TParams extends PromptInputParams = Record<string, never>>(
  params: TParams,
  opts: BasePromptOptions = {}
): Promise<Resolve<InputToOutputParamValues<TParams>>> => {
  // ...
};
prompt.confirm = confirm;
prompt.integer = integer;
// ...
```

ericliuche commented:
@colinking ready now!

colinking commented on packages/airplane/internal/execute.test.ts:
```diff
@@ -35,7 +35,7 @@ describe("execute", () => {
         paramValues: {
           name: "colin",
           file: "testid",
-          date: "2022-10-31T00:00:00-07:00",
```
This looks wrong. We purposefully serialize date types in the current user's timezone so that we don't get "off-by-one" bugs where switching to UTC changes the day.

colinking commented on packages/airplane/internal/params/params.ts:
```diff
@@ -25,6 +25,28 @@ export type Param = { type: ParamType } & (
   | ConfigVarParam
 );
 
+// TODO: support configvars and uploads
```
Can you add a Linear? Easier to track.

colinking commented on packages/airplane/internal/params/params_input.ts:
```diff
@@ -54,6 +62,13 @@ type InputParam = ParamType | Param;
  */
 export type InputParams = Record<string, InputParam>;
 
+/**
+ * PromptInputParams are a set of parameters keyed by parameter slug available to prompts.
+ *
+ * The order of parameters within the object is retained for ordering parameters in form UIs.
+ */
```
```suggestion
/**
 * PromptInputParams are the subset of InputParams that are available to prompts.
 *
 * Currently config variables and files are not supported by prompts. Once those are supported,
 * we can remove this in favor of using InputParams directly.
 */
```

colinking commented on packages/airplane/internal/params/params_output.test.ts:
```diff
@@ -0,0 +1,87 @@
+import { deserializeParamValues } from "./params_output";
+
+describe("deserializeParamValues", () => {
```
nice

colinking commented on packages/workflow-runtime/internal/index.ts:
```diff
@@ -92,20 +93,22 @@ export const runtime: RuntimeInterface = {
   },
 
-  prompt: async (prompt: PromptConfig): Promise<SerializedParamValues> => {
+  prompt: async <TParams extends InputParams = Record<string, never>>(
```
I find it easier to avoid defaults unless there's a good reason for them (e.g. a type exported to end-users that won't get inferred).

colinking commented on packages/workflow-runtime/internal/index.ts:
```diff
@@ -92,20 +93,22 @@ export const runtime: RuntimeInterface = {
 
-  prompt: async (prompt: PromptConfig): Promise<SerializedParamValues> => {
+  prompt: async <TParams extends InputParams = Record<string, never>>(
+    prompt: PromptConfig
```
nit: rename to `CreatePromptRequest`. Better conveys this is an API type. I thought it was the user-facing type at first.

colinking commented on packages/airplane/internal/prompt/index.ts:
```diff
@@ -27,7 +31,7 @@ export const text = async (name: string, opts: PromptOptions<string> = {}): Prom
     cancelText: opts.cancelText,
     confirmText: opts.confirmText,
   });
-  return values[param.slug] as string;
```
:awyis: I love this

colinking commented on packages/airplane/internal/prompt/index.ts:
```diff
@@ -111,10 +115,8 @@ export const date = async (
+  return new Date(values[param.slug]);
 };
 
 export const datetime = async (
```
Minor, but would it make sense to rewrite all these methods to just call `.prompt()`? e.g. you wouldn't need to do the conversion to output values.

colinking replied:
Doing so should fix https://linear.app/airplane/issue/TSK-465/prompt-output-should-be-undefined-when-optional for these methods

ericliuche replied:
hmm would that require us having to write a template that converts `PromptOptions` to `InputParams`?

colinking reviewed (commented):
Oof there's a lot here 😄 First pass of comments.

colinking commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+  optional: boolean
+): Array<{ label: string; value: SerializedParamValue }> => {
+  if (options.length === 0 && !optional) {
+    throw new Error(`The required prompt parameter, "${slug}", has no options to select from.`);
```
This function isn't directly related to prompts.

```suggestion
    throw new Error(`The required parameter, "${slug}", has no options to select from.`);
```

colinking replied:
Hmm, does this match our API semantics? I would almost assume if a user passes an empty array of options, we would ignore it.

ericliuche replied:
yup you're right. should we add the check to parameter verification?

<img width="484" alt="Screen Shot 2022-11-28 at 3 40 56 PM" src="https://user-images.githubusercontent.com/11494140/204376548-b248f6d3-3264-4d33-a7d4-7661fb7d744c.png">

colinking replied:
We apparently do this in our UI:

<img width="784" alt="CleanShot 2022-11-28 at 18 05 42" src="https://user-images.githubusercontent.com/2907397/204399442-f0583f58-8aaf-4126-983b-bbe340ee675d.png">

Returning an error from the API SGTM. Can you add a Linear for that?

ericliuche replied:
https://linear.app/airplane/issue/TSK-1193/add-parameter-verification-check-that-empty-options-is-not-allowed

colinking commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+  return options.map((option) => ({
+    label: optionToLabel(option),
+    value: serializeParamValue(
+      typeof option === "object" && "value" in option ? option.value : option
```
I'd add a helper for this:

```
const isOptionObject = <TOption>(
  opt: TOption | SelectOption<TOption>
): opt is SelectOption<TOption> => {
  return typeof opt === "object" && opt != null && "value" in opt;
};
```

FYI I noticed `label` is required in `SelectOption` but should be optional. Can you fix?

colinking commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+    if (option instanceof Date) {
+      // TODO: Currently we incorrectly serialize `date` parameters the same as `datetime`.
+      //       `date` parameters need to instead be formatted as `YYYY-MM-DD`.
+      return option.toISOString();
```
Same here as above -- let's continue to use `toISOString(option)` so we retain the time zone when serialized for dates.

colinking commented on packages/airplane/internal/prompt/index.ts:
```diff
@@ -209,3 +209,18 @@ export const select = async <TOption = unknown>(
   return options[i];
 };
+
+export const prompt = async <TParams extends PromptInputParams = Record<string, never>>(
```
You'll want to remove this `Record<string, never>` default as it breaks intellisense:

before:
<img width="840" alt="CleanShot 2022-11-23 at 13 15 28" src="https://user-images.githubusercontent.com/2907397/203619589-be610aac-f6ed-49bd-bcfe-165c98e944d7.png">

after:

<img width="882" alt="CleanShot 2022-11-23 at 13 15 47" src="https://user-images.githubusercontent.com/2907397/203619594-5f3ad561-443d-4c05-8a36-54a75e909903.png">

colinking reviewed (changes requested):
LGTM, apart from some minor comments below. Is the plan to keep the old methods, or do we want to deprecate them? What's the plan around adding the alternative array notation?

colinking commented:
@ericliuche one more ask -- can you add JSDoc comments to the prompt methods, similar to what I did in this PR? https://github.com/airplanedev/js-sdk/pull/208

ericliuche commented on packages/airplane/internal/params/params_serialized.ts:
```diff
@@ -61,7 +68,106 @@ export const serializeParamValue = (paramValue: InputParamValue): SerializedPara
+      const optional = param.required === false ? true : false;
+      serializedParams.push({
+        slug: slug,
+        name: param.name ? param.name : slug,
```
i'll add a task to do this in a later change, we already do it for the python sdk
`name=param_config.name or inflection.humanize(default_slug),`

ericliuche commented:
`Is the plan to keep the old methods, or do we want to deprecate them?`
i think we should deprecate the old methods, should i just mark them as deprecated here?

`What's the plan around adding the alternative array notation?`
we can do that in a follow up change, i'll add a task for it

colinking commented:
> i think we should deprecate the old methods, should i just mark them as deprecated here?

Sweet, use this annotation for that and VSCode will mark usage with a strikethrough: https://jsdoc.app/tags-deprecated.html

ericliuche commented:
`JSDoc comments to the prompt methods` 

@colinking i just added docs to the new `prompt` method I added since the other methods are deprecated now. is that fine?

colinking commented on packages/airplane/internal/prompt/index.ts:
```diff
@@ -17,128 +24,149 @@ export const confirm = async (opts: BasePromptOptions = {}): Promise<void> => {
 /**
  * Prompts a user to select one item from a list of options.
  *
+ * @deprecated since v0.2.23, use `prompt()` instead
```
This method's actually a convenience method (given an arbitrary list of elements, you can create a select parameter from it), so there's actually no direct replacement with `prompt`. Is there a way we can keep it or something similar around?

ericliuche replied:
ah good point, will just remove the deprecated tag for now

colinking commented on packages/airplane/internal/prompt/prompt.test-d.ts:
```diff
@@ -0,0 +1,95 @@
+const _prompt = async () => {
+  // Prompt with no parameters without arg.
+  {
+    const _myPrompt = prompt({});
```
add twoslash check

colinking commented on packages/airplane/internal/params/params_serialized.test.ts:
```diff
@@ -0,0 +1,623 @@
+      },
+      e: {
+        type: "shorttext",
+        regex: "$.*^",
```
```suggestion
        regex: "^.*$",
```
